### ISSUE TITLE
Graph classification (inference)

### ISSUE THREAD

**Issue by mhillebrand** [graph4ai/graph4nlp#458]
I just tried `inference.py` in the `develop` branch for sentence (dependency graph) classification , which is great for classifying single sentences here and there, but with large amounts of sentences to classify, it's quite slow. For each input sentence, a separate connection to the Stanford CoreNLP server is made—even if I set `batch_size` to something greater than one. This certainly isn't the case when training a graph classifier; that script did dependency parsing in large batches and also used multiple/concurrent CoreNLP threads, I believe.

https://github.com/graph4ai/graph4nlp/blob/df6c31bc0a00fc57bc8148a3578a229640069025/graph4nlp/pytorch/inference_wrapper/base.py#L105

```
    def preprocess(self, raw_contents: list):
        processed_data_items = []
        use_ie = self.graph_name == "ie"  # hard code
        for raw_sentence in raw_contents:
            if self.data_item_class == DoubleText2TextDataItem:
                assert isinstance(
                    raw_sentence, (tuple, list)
                ), "Expect two inputs for DoubleText2TextDataItem, only one found."
                data_item = self.data_item_class(
                    input_text=raw_sentence[0],
                    input_text2=raw_sentence[1],
                    output_text=None,
                    tokenizer=self.tokenizer,
                    share_vocab=self.share_vocab,
                )
            elif self.data_item_class == Text2TreeDataItem:
                data_item = self.data_item_class(
                    input_text=raw_sentence,
                    output_text=None,
                    output_tree=None,
                    tokenizer=self.tokenizer,
                )

            else:
                data_item = self.data_item_class(input_text=raw_sentence, tokenizer=self.tokenizer)

            data_item = self.dataset.process_data_items(data_items=[data_item])
            data_item = self.dataset._vectorize_one_dataitem(
                data_item[0], self.vocab_model, use_ie=use_ie
            )
            processed_data_items.append(data_item)
        return processed_data_items
```

I took a glance at `inference_advance.py`, but it deals with target tensors and evaluating accuracy (and has no `predict` method), so I assumed that it's still a work in progress, or is there something I'm misunderstanding about this script?

**Comment by AlanSwift** [comment]
Thanks for your interest.
Yes, currently it is in progress. But there is some misunderstanding as well.
The ``inference_advance.py`` and ``inference.py`` have the same effect but for different purposes actually.
They can all get the predicted results (I mean both the logits and the string-like outputs).

The ``inference.py``adopts the ``inference_wrapper`` which can directly handle the raw inputs, which is convenient for the user to load the pre-trained weights and do the inference. It aims to handle the immediate inputs. Thus the ``graph4nlp`` should parse the raw inputs and then do the forward calculation.

The ``inference_advance.py`` serves as an advanced practice for the inference. Basically, it is used for the large-scale dataset. It will do the parsing and graph construction in preprocessing and store the results in the disk to speed up the inference speed. However, users should be familiar with the library, especially the preprocessing, which is not so friendly.

In short, the ``inference.py`` will do the parsing on the fly. The constructed graph results will not be saved. And currently, it is a preview version and doesn't support multi-processing. So pros: user-friendly; cons: slow.
And the ``inference_advance.py`` is the opposite. pros: quick; cons: it is not user-friendly in some way

However, your advice of adding multi-processing support for ``inference_wrapper`` is a good idea. We will support it in the next version (or maybe the second version according to our plan).

**Comment by AlanSwift** [comment]
The multiprocess support is done in pr #550 .
